### PR TITLE
Avoid zeroing a 1kb stack buffer on every call to `std::sys::windows::fill_utf16_buf`

### PR DESCRIPTION
**Pull request by thomcc**
I've also tried to be slightly more careful about integer overflows, although in practice this is likely still not handled ideally.

r? @ChrisDenton

### CODE REVIEW

**Comment by thomcc** [comment]
FYI: I wrote this on macOS, so probably don't land this without a try run first (it *should* be fine but for all I know I got some logic backwards or something).

**Comment by ChrisDenton** [comment]
Well it doesn't look like there's anything in the try queue atm so we might as well use it for something...

@bors try

**Comment by Mark-Simulacrum** [comment]
Try builds are Linux only dist builds without extra work so it probably won't show anything interesting.

**Comment by thomcc** [comment]
> Try builds are Linux only dist builds without extra work so it probably won't show anything interesting.

Nards. I was pretty sure I'd seen them run the windows CI tests before.

**Comment by ChrisDenton** [comment]
Oh! That's annoying. If it helps I can test locally when I have a mo.

**Comment by Mark-Simulacrum** [comment]
https://github.com/rust-lang/rust/blob/master/.github/workflows/ci.yml#L7 has the file you need to edit and rerun the expander to regenerate the actual configuration. Mostly just a matter of copying a windows job to the try: section.

**Comment by klensy** [comment]
There few similar places like that
https://github.com/rust-lang/rust/blob/8064a495086c2e63c0ef77e8e82fe3b9b5dc535f/library/std/src/sys/windows/stdio.rs#L60
https://github.com/rust-lang/rust/blob/8064a495086c2e63c0ef77e8e82fe3b9b5dc535f/library/std/src/sys/windows/stdio.rs#L171-L173

**Comment by ChrisDenton** [comment]
Ok I've tested this locally and it looks great so let's get this merged.

@bors r+ rollup

File: library/std/src/sys/windows/mod.rs [hunk 1]
```diff
@@ -244,7 +250,9 @@ where
                 // Therefore k never equals n.
                 unreachable!();
             } else {
-                return Ok(f2(&buf[..k]));
+                // Safety: First `k` bytes are initialized.
```
Comment (5225225) [inline library/std/src/sys/windows/mod.rs:253]: It's the first `k` `u16`s that are initialized, not bytes.

**Pull request merged**